### ISSUE TITLE
nn.TransformerEncoder - all nan values issues when src_key_padding_mask provided

### ISSUE THREAD

**Issue by semirelezovikj** [pytorch/pytorch#64525]
I have a use case where I am dealing with sequences of variables lengths. I am looking to make use of nn.TransformerEncoder, but for some reason if a given sequence is of a length < max_length of sequence, all values result in nan in the forward pass.

Explanation of the code
I am including the fully reproducible code below.

In the example below, batch_size is 2, the maximum sequence length is 3, and I provide two sequences: one with 2 tokens and the other one with 3 tokens. Naturally, the sequence with 2 tokens needs to be padded in order to be fed to nn.TransformerEncoder.

In order to do this, I need to provide src_key_padding_mask of shape (N, S) where N is the batch_size and S is the sequence_length, in order to provide per-batch padding mask. For the sequence of length 2, a value of True needs to be provided for the position that is being padded - this means that for this particular sequence the padding values will be [False, False, True]

To my surprise, providing the src_key_padding_mask results with all nan values for the entire sequence where the sequence length < max_sequence_length.

Note: I have looked through all the other issues where other users were getting all nan values, but their scenarios were due to “user error”, where they provided sequences where all tokens are padded. This is not the case with my scenario.

```
from torch.nn.utils.rnn import pad_sequence
import torch
import torch.nn as nn
ego_encoder_layer = nn.TransformerEncoderLayer(d_model=2048, nhead=16, activation='gelu')
ego_transformer_encoder = nn.TransformerEncoder(ego_encoder_layer, num_layers=6)

sequences = []
batch_size = 2
src_padding_mask = torch.zeros((batch_size, 3)).type(torch.bool)

sequences.append(torch.rand(2, 2048))
sequences.append(torch.rand(3, 2048))
seq_lengths = [2, 3]

ego_seq2 = pad_sequence(sequences, padding_value=float('-inf'))
print(ego_seq2.shape)

ego_transformer_features = ego_transformer_encoder(ego_seq2, src_key_padding_mask=src_padding_mask)
print(ego_transformer_features.shape)




max_seq_length = 3
for ind, sl in enumerate(seq_lengths):
    for j in range(sl, max_seq_length):
        src_padding_mask[ind, j] = True # True means this is a padding token - skip

print('- seq_lengths, src_padding_mask -')
print(seq_lengths, src_padding_mask)

print('transformer features')
print(ego_transformer_features.shape)
ego_transformer_features[:, :, :5]
```

Output: 

```
torch.Size([3, 2, 2048])
torch.Size([3, 2, 2048])
- seq_lengths, src_padding_mask -
[2, 3] tensor([[False, False,  True],
        [False, False, False]])
transformer features
torch.Size([3, 2, 2048])
tensor([[[    nan,     nan,     nan,     nan,     nan],
         [ 1.0203, -0.1215,  1.7757,  2.8037,  0.2745]],

        [[    nan,     nan,     nan,     nan,     nan],
         [ 1.4554,  0.1737,  1.9316,  2.8318,  1.0204]],

        [[    nan,     nan,     nan,     nan,     nan],
         [ 0.4414,  1.2467,  0.6188,  2.2753,  0.2954]]],
       grad_fn=<SliceBackward>)
```

**Comment by timoklein** [comment]
Hey,
I think I can help since I ran into a similar issue for a project of mine. The problem is related to using infinity as padding value in this line
`ego_seq2 = pad_sequence(sequences, padding_value=float('-inf'))`
It will work if you use a large constant, e.g. `1e10` or `-1e10` instead of `float('-inf')`.

**Comment by semirelezovikj** [comment]
@timoklein 

I am aware that it will work with `1e10` (output below) - sequence 0 has 2 elements. You can see after the forward pass that the element that should be attended (True in the src_key_padding_mask) still has none-0 elements. I would expect that the value would be all zero or nan for the positions that are not being attended.

**Output with 1e-10 as padding token**:
```
torch.Size([3, 2, 2048])
torch.Size([3, 2, 2048])
- seq_lengths, src_padding_mask -
[2, 3] tensor([[False, False,  True],
        [False, False, False]])
transformer features
torch.Size([3, 2, 2048])
tensor([[[ 0.4437, -1.0837,  0.7559, -1.6229,  0.4586],
         [-0.6258,  0.0834, -0.5848,  0.5980, -0.6840]],

        [[ 0.4486, -1.9860,  0.3684, -0.7162,  0.7047],
         [-0.8633, -0.3746, -1.1330,  0.2826, -0.7749]],

        [[ 0.4362, -0.7720,  0.8519,  0.1682,  0.9219],
         [-1.1458, -0.4938, -0.1952, -1.1702, -0.2074]]],
       grad_fn=<SliceBackward>)
```

**Comment by semirelezovikj** [comment]
It's getting curiouser and curiouser - I set a manual_seed to torch on the beginning. 

Setting the padding_token to `1e-10` has significantly different results than `1e10` on the tokens that should be attended to. I am surprised that the choice of padding value for tokens that should be skipped affects the output of the tokens that should be skipped.



```
SEED=12345
torch.manual_seed(SEED)
torch.cuda.manual_seed(SEED)
torch.backends.cudnn.deterministic = True
```

**Output with 1e-10 as padding token:**

```
torch.Size([3, 2, 2048])
torch.Size([3, 2, 2048])
- seq_lengths, src_padding_mask -
[2, 3] tensor([[False, False,  True],
        [False, False, False]])
transformer features
torch.Size([3, 2, 2048])
tensor([[[-0.3584,  0.5694,  0.6457,  1.5897,  0.5229],
         [ 1.9117,  1.1189,  0.1857, -0.0905, -0.8727]],

        [[ 0.9486,  0.3187,  0.3235,  1.0732, -0.1852],
         [ 1.4643,  1.5484, -0.4661,  1.1593, -1.2341]],

        [[ 0.6822,  0.8850,  0.5383,  0.8657,  0.4569],
         [ 1.5199,  0.5866, -0.0431,  0.8116, -1.7867]]],
       grad_fn=<SliceBackward>)
```

**Output with 1e10 as padding token:**

```
torch.Size([3, 2, 2048])
torch.Size([3, 2, 2048])
- seq_lengths, src_padding_mask -
[2, 3] tensor([[False, False,  True],
        [False, False, False]])
transformer features
torch.Size([3, 2, 2048])
tensor([[[ 1.1449,  2.8843,  0.7907,  0.5957,  0.6778],
         [ 1.9117,  1.1189,  0.1857, -0.0905, -0.8727]],

        [[ 1.2219,  2.6348,  0.4374, -0.0962,  0.3763],
         [ 1.4643,  1.5484, -0.4661,  1.1593, -1.2341]],

        [[ 1.1898,  3.4397,  0.6832,  0.1120,  0.7404],
         [ 1.5199,  0.5866, -0.0431,  0.8116, -1.7867]]],
       grad_fn=<SliceBackward>)
```

**Comment by timoklein** [comment]
> @timoklein
> 
> I am aware that it will work with `1e10` (output below) - sequence 0 has 2 elements and sequence 1 has 3 elements. You can see after the forward pass that the element that should **not** be attended (True in the src_key_padding_mask) still has none-0 elements. I would expect that the value would be all zero or nan for the positions that are not being attended.
> I am worried that the `src_key_padding_mask` is incorrectly coded to still attend to the places marked "not for attention".
> 
> **Output with 1e-10 as padding token**:
> 
> ```
> torch.Size([3, 2, 2048])
> torch.Size([3, 2, 2048])
> - seq_lengths, src_padding_mask -
> [2, 3] tensor([[False, False,  True],
>         [False, False, False]])
> transformer features
> torch.Size([3, 2, 2048])
> tensor([[[ 0.4437, -1.0837,  0.7559, -1.6229,  0.4586],
>          [-0.6258,  0.0834, -0.5848,  0.5980, -0.6840]],
> 
>         [[ 0.4486, -1.9860,  0.3684, -0.7162,  0.7047],
>          [-0.8633, -0.3746, -1.1330,  0.2826, -0.7749]],
> 
>         [[ 0.4362, -0.7720,  0.8519,  0.1682,  0.9219],
>          [-1.1458, -0.4938, -0.1952, -1.1702, -0.2074]]],
>        grad_fn=<SliceBackward>)
> ```

Off the top of my head: After each Multi-Head Attention layer the Transformer encoder uses a feedforward layer. You can see this e.g. [here](https://jalammar.github.io/illustrated-transformer/). It's most likely also in the pytorch implementation of an encoder layer. Thus you will not get 0s after masking when testing with an encoder layer. You need to use `nn.MultiheadAttention` if you want to see the effect of your masking.

**Comment by semirelezovikj** [comment]
@timoklein How about my second post?

"Setting the padding_token to 1e-10 has significantly different results than 1e10 on the tokens that should be attended to. I am surprised that the choice of padding value for tokens that should be skipped affects the output of the tokens that should NOT be skipped."

**Comment by timoklein** [comment]
Maybe it is because you first compute the features in this line
`ego_transformer_features = ego_transformer_encoder(ego_seq2, src_key_padding_mask=src_padding_mask)`
with this mask you previously generated
`src_padding_mask = torch.zeros((batch_size, 3)).type(torch.bool)` which does not mask anything.
Only after you computed the features you actually set a value in your mask to **True** in these lines
```
max_seq_length = 3
for ind, sl in enumerate(seq_lengths):
    for j in range(sl, max_seq_length):
        src_padding_mask[ind, j] = True # True means this is a padding token - skip
```
Since your mask does not mask anything when you do the computation, the Transformer attends to the pad tokens as well. If you change the order it should work.

**Comment by semirelezovikj** [comment]
Oops - my bad. That was an issue with the way I tried to recreate my problem.

**Comment by jbschlosser** [comment]
@semirelezovikj Just to confirm, once you correctly order the logic to build the mask before using it, you don't see any differences between `padding=1e10` and `padding=1e-10`?

**Comment by cpuhrsch** [comment]
@pearu - this should be a use case for [masked_softmax](https://github.com/pytorch/rfcs/pull/27), since it allows to actually mask out values and prevents interactions between inf and -inf in this specific case.

**Comment by jbschlosser** [comment]
@reedts Have you tried padding with finite values at the masked positions as mentioned above (e.g. `0.` should work)?

I agree that the values at the masked positions should ideally have no impact on the output. The reason they do now is because masking is only applied after projection, and projecting with infinite values in the inputs causes NaNs.

Note that masked position values will not be attended to for outputs at unmasked positions, so avoiding the NaNs is enough to get the desired behavior.

@cpuhrsch is working on proper masking semantics for tensors, which should make the issue solvable so that the values at masked-out positions truly do not matter at all.
